### PR TITLE
fix(datastore): write nil interface as a nil property with Save()

### LINKED ISSUE #2399: datastore: nil interface{} panics on Save
**Client**

Datastore

**Environment**

Emulator

**Code**

https://github.com/googleapis/google-cloud-go/blob/e2b907844a20f22de44a70de164ed53dced5b33a/datastore/save.go#L67-L71

**Expected behavior**

Uninitialized `interface{}` value (`nil`) correctly written as nil property value, such as the case with `reflect.Ptr`:
https://github.com/googleapis/google-cloud-go/blob/e2b907844a20f22de44a70de164ed53dced5b33a/datastore/save.go#L79-L85

**Actual behavior**

The `v` variable replacement on L70 will yield a `reflect.Value` of type `nil`. When the method is invoked recursively on L71, the call to `v.Interface()` will panic:
https://github.com/googleapis/google-cloud-go/blob/e2b907844a20f22de44a70de164ed53dced5b33a/datastore/save.go#L53

**Screenshots**
image

**Additional context**

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/ermik/Library/Caches/go-build"
GOENV="/Users/ermik/Library/Application Support/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/ermik/Projects/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/Cellar/go/1.13.4/libexec"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/Cellar/go/1.13.4/libexec/pkg/tool/darwin_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="/Users/ermik/Projects/[projec-path]/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/wp/bpblv6xj33n17_8pr7h8k1t40000gn/T/go-build792484503=/tmp/go-build -gno-record-gcc-switches -fno-common"
</pre></details>

Struct being saved: [cayleygraph/quad.Quad](https://github.com/cayleygraph/quad/blob/2a07fc872617fd9a1c28ad38ece065ca984f798b/quad.go#L110-L115) — `Label` is nil.

Module version from `go.mod` — `cloud.google.com/go/datastore v1.1.0`

### PR DESCRIPTION
**Pull request by IlyaFaer**
Closes #2399

### CODE REVIEW

File: datastore/save_test.go [hunk 1]
```diff
@@ -333,6 +333,19 @@ func TestSaveFieldsWithInterface(t *testing.T) {
...
+				Value interface{}
+				key   interface{}
+			}{
+				Value: nil,
+				key:   "key1",
+			},
+			want: []Property{
+				{Name: "Value", Value: nil},
+			},
+		},
```
Comment (IlyaFaer) [inline datastore/save_test.go:348]: I've been able to reproduce the error with such a test case. See the screenshot for traceback:

Безымянный

File: datastore/save.go [hunk 2]
```diff
@@ -67,8 +67,13 @@ func reflectFieldSave(props *[]Property, p Property, name string, opts saveOpts,
...
 			// See issue https://github.com/googleapis/google-cloud-go/issues/1474.
-			v = v.Elem()
-			return reflectFieldSave(props, p, name, opts, v)
+			if v.IsNil() {
+				// Nil pointer becomes a nil property value (unless omitempty, handled above).
+				p.Value = nil
+				*props = append(*props, p)
+				return nil
+			}
+			return reflectFieldSave(props, p, name, opts, v.Elem())
```
Comment (IlyaFaer) [inline datastore/save.go:76]: Small piece of code, so I didn't move it into a separate function

File: datastore/save.go [hunk 3]
```diff
@@ -67,8 +67,13 @@ func reflectFieldSave(props *[]Property, p Property, name string, opts saveOpts,
 		case reflect.Interface:
 			// Extract the interface's underlying value and then retry the save.
 			// See issue https://github.com/googleapis/google-cloud-go/issues/1474.
-			v = v.Elem()
-			return reflectFieldSave(props, p, name, opts, v)
+			if v.IsNil() {
+				// Nil pointer becomes a nil property value (unless omitempty, handled above).
```
Comment (tritone) [inline datastore/save.go:71]: Where is omitEmpty handled? Comment could be clearer.

Comment (IlyaFaer) [reply datastore/save.go]: Agree. Probably, "embracing call" will sound better

File: datastore/save.go [hunk 4]
```diff
@@ -67,8 +67,14 @@ func reflectFieldSave(props *[]Property, p Property, name string, opts saveOpts,
 		case reflect.Interface:
 			// Extract the interface's underlying value and then retry the save.
 			// See issue https://github.com/googleapis/google-cloud-go/issues/1474.
-			v = v.Elem()
-			return reflectFieldSave(props, p, name, opts, v)
+			if v.IsNil() {
+				// Nil interface becomes a nil property value (unless omitempty,
+				// handled by the embracing saveStructProperty call).
```
Comment (tritone) [inline datastore/save.go:72]: How about this:
```
// Nil interface becomes a nil property value (unless omitEmpty is true,
// which is handled by the caller).
```

Please change L88 to say the same thing for a nil pointer as well. Thanks!

Comment (IlyaFaer) [reply datastore/save.go]: No problem, pushed

**Comment by ermik** [comment]
👏

**Pull request merged**